usecase/syncstrategy/window.go

```diff
@@ -46,42 +45,33 @@ func (window *Window) Sync(
 	})
 	logger.Debug("spawning goroutines for sync block workers")
 
+	workersWaitGroup := sync.WaitGroup{}
+	workersWaitGroup.Add(int(endHeight - beginHeight + 1))
+	commandWindowMutex := sync.Mutex{}
+
+	commandWindow := newUnsafeCommandWindow(beginHeight, endHeight)
+
 	for height := beginHeight; height <= endHeight; height += 1 {
 		go func(height int64) {
 			commands, err := worker(height)
+			result := workResult{height, commands, nil}
+
 			if err != nil {
-				workResultCh <- workResult{height, nil, err}
+				logger.Errorf("received error from sync block worker #%d: %v", height, err)
+				result = workResult{height, nil, err}
 			}
 
-			workResultCh <- workResult{height, commands, nil}
+			commandWindowMutex.Lock()
+			commandWindow.Put(result.height, result.commands)
+			commandWindowMutex.Unlock()
+
+			workersWaitGroup.Done()
```

Comment:
One critque I will put is the use of Mutex here while in the original design it can be achieved using channel in a non-blocking manner to the goroutines. (but we probably should increase the channel size).

Reply:
Also we probably need some error handling here and use `defer` to Unlock the window.

Reply:
I still think the use of channel is better in terms of error handling and running things in parallel in general.

Although in this case since the critical section is at the end of goroutine so it may not hurt the correctness.

Reply:
Got your point. Actually, the original design was in a blocking manner already, we can see similar behaviour in here https://play.golang.org/p/BZArLAwhlym

As we can see the go routines is finally blocking each other since
```
for {
	temp := <- ch
	...
)
```
is the blocking statement

the same as original design
```
for {
	result := <-workResultCh
	...
)
```

Reply:
Checked slice assignment is thread safe, we can improve from original design by removing the above mutex lock.

Reply:
> Actually, the original design was in a blocking manner already, we can see similar behaviour in here

Yes, this is correct. This could be resolved by allocating a larger size channel.

Default channel: https://play.golang.org/p/oeTbXE0Ry4H
Explicitly sized channel: https://play.golang.org/p/io4EeKOYj4W

Reply:
> Checked slice assignment (using index) is thread safe, we can improve from original design by removing the above mutex lock.

Sure, if that's the case then the Window is not "unsafe" anymore.

Reply:
Updated, workers should be running in parallel